src/native/libs/System.Security.Cryptography.Native.Android/net/dot/android/crypto/PalPbkdf2.java

```diff
@@ -0,0 +1,101 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+package net.dot.android.crypto;
+
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import javax.crypto.Mac;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+
+public final class PalPbkdf2 {
+    private static final int ERROR_UNSUPPORTED_ALGORITHM = -1;
+    private static final int SUCCESS = 1;
+
+    public static int pbkdf2OneShot(String algorithmName, byte[] password, byte[] salt, int iterations, byte[] destination)
+        throws ShortBufferException, InvalidKeyException {
+
+        // We do not ever expect a ShortBufferException to ever get thrown since the buffer destination length is always
+        // checked. Let it go through the checked exception and JNI will handle it as a generic failure.
+        // InvalidKeyException should not throw except the the case of an empty key, which we already handle. Let JNI
+        // handle it as a generic failure.
+
+        // We use a custom implementation of PBKDF2 instead of the one provided by the Android platform for two reasons:
+        // The first is that Android only added support for PBKDF2 + SHA-2 family of agorithms in API level 26, and we
+        // need to support SHA-2 prior to that.
+        // The second is that PBEKeySpec only supports char-based passwords, whereas .NET supports arbitrary byte keys.
+
+        assert algorithmName != null;
+        assert password != null;
+        assert salt != null;
+        assert iterations > 0;
+        assert destination != null;
```

Comment:
Can I assume that Java assertions are enabled on debug builds?

Reply:
Hm. Actually no. Java does not really have the concept of a "debug" and "release" build, and we aren't building with `-enableassertions`. I will change this to be an exception.

Reply:
I think my question was "are we building with -enableassertions in our debug builds?".

For the OpenSSL/Linux shim, older code was all runtime checks, but since we have assertions in debug builds I think we've generally gone for saying assert is fine for "you're holding it wrong" (e.g. "we only have one caller, why would they have passed null for the destination?!")... so perhaps the better move is to just specify `-enableassertions` for the Java components for our debug builds...

Unless what I'm missing here is that the java things don't actually build in `./build libs` but some later step where we've lost the notion of if we were a debug or release build of the libraries slice.

Reply:
> I think my question was "are we building with -enableassertions in our debug builds?".

We never pass in `-enableassertions` regardless of our build configuration. Seems that Java assertions are not super popular in the Java ecosystem. "Just throw an exception".